runtime/nsd_gtls.c

```diff
@@ -1592,13 +1592,16 @@ Rcv(nsd_t *pNsd, uchar *pBuf, ssize_t *pLenBuf)
 	*pLenBuf = iBytesCopy;
 
 finalize_it:
-	if (iRet != RS_RET_OK) {
-		/* in this case, we also need to free the receive buffer, if we
+	if (iRet != RS_RET_OK && 
+		iRet != RS_RET_RETRY) {
+		/* We need to free the receive buffer in error error case unless a retry is wanted. , if we
 		 * allocated one. -- rgerhards, 2008-12-03 -- moved here by alorbach, 2015-12-01
 		 */
 		*pLenBuf = 0;
-		free(pThis->pszRcvBuf);
-		pThis->pszRcvBuf = NULL;
+		if(pThis->pszRcvBuf != NULL) {
```

Comment:
This is not necessary, free(NULL) is defined to be a valid (no-)operation. I just point this out in case you consider this an important part of the fix. Else I would suggest to remove it again (for clarity).